src/Lean/Meta/MkIffOfInductiveProp.lean

```diff
@@ -0,0 +1,380 @@
+          let (_, mv3) ← mv2.revert fvars'.toArray
+          let mv4 ← fvars'.foldlM (fun mv _ ↦ do
+            let ⟨fv, mv'⟩ ← mv.intro1
+            let #[res] ← mv'.cases fv | throwError "expected one case subgoal"
+            return res.mvarId) mv3
+          pure (mv4, fvars, numHEqs)
+        mvar'.withContext do
+          let fvarIds := (← getLCtx).getFVarIds.toList
+          let gs := fvarIds.take gs.length
+          let hs := fvarIds.extract (fvarIds.length - (n + numHEqs)) (fvarIds.length - numHEqs)
+          let m := gs.map some ++ listBoolMerge bs hs
+          let args ← m.mapM fun a ↦
+            match a with
+            | some v => pure (mkFVar v)
+            | none => mkFreshExprMVar none
+          let c ← mkConstWithLevelParams constr_name
+          let e := mkAppN c args.toArray
+          let t ← inferType e
+          let mt ← mvar'.getType
+          let _ ← isDefEq t mt -- infer values for those mvars we just made
+          mvar'.assign e
+
+/--
+  Generates existential form of a prop-valued inductive type and proves the equivalence.
+-/
+private def mkIffOfInductivePropImpl (inductVal : InductiveVal) (rel : Name) : MetaM Unit := do
+  let constrs := inductVal.ctors
+  let params := inductVal.numParams
+  let type := inductVal.type
+
+  let univNames := inductVal.levelParams
+  let univs := univNames.map mkLevelParam
+  /- we use these names for our universe parameters, maybe we should construct a copy of them
+  using `uniq_name` -/
+
+  let (thmTy, shape, existential) ← Meta.forallTelescopeReducing type fun fvars ty ↦ do
+    if !ty.isProp then throwError "mk_iff only applies to prop-valued declarations"
+    let lhs := mkAppN (mkConst inductVal.name univs) fvars
+    let fvars' := fvars.toList
+    let shape_rhss ← constrs.mapM (constrToProp univs (fvars'.take params) (fvars'.drop params))
+    let (shape, rhss) := shape_rhss.unzip
+    let existential := mkOrList rhss
+    let existential ← mkLambdaFVars fvars existential
+    pure (← mkForallFVars fvars (mkApp2 (mkConst `Iff) lhs (mkOrList rhss)), shape, existential)
+
+  trace[Meta.MkIffOfInductiveProp] "Existential form is: {existential}"
+  trace[Meta.MkIffOfInductiveProp] "The type of proof of equivalence: {thmTy}"
+
+  let mvar ← mkFreshExprMVar (some thmTy)
+  let mvarId := mvar.mvarId!
+  let (fvars, mvarId') ← mvarId.intros
+  let [mp, mpr] ← mvarId'.apply (mkConst `Iff.intro) | throwError "failed to split goal"
+
+  toCases mp shape
+  let ⟨mprFvar, mpr'⟩ ← mpr.intro1
+  toInductive mpr' constrs ((fvars.toList.take params).map .fvar) shape mprFvar
+
+  let proof ← instantiateMVars mvar
+  addDecl <| .thmDecl {
+    name := rel
+    levelParams := univNames
+    type := thmTy
+    value := proof
+  }
+
+  addDecl <| .defnDecl (←mkDefinitionValInferringUnsafe
+    (inductVal.name ++ `existential) inductVal.levelParams
+    (←inferType existential) existential .opaque)
```

Comment:
If you are using `mkDefinitionValInferringUnsafe` I assume you are expecting to handle `unsafe` definitions, then maybe also use (the relatively new) `mkThmOrUnsafeDef` above.

Do you have a test for `unsafe coinductive` (not yet it seems)?